security/keys/encrypted-keys/encrypted.c

```diff
@@ -956,6 +1042,77 @@ static long encrypted_read(const struct key *key, char *buffer,
 	return ret;
 }
 
+/*
+ * user_encrypted_read - format and copy out the encrypted data
+ *
+ * The resulting datablob format is:
+ * <master-key name> <decrypted data length> <encrypted iv> <encrypted data>
+ *
+ * Unless the format is 'unwrappable':
+ * <master-key name> <decrypted data length> <encrypted iv> <encrypted data> <decrypted data>
+ *
+ * On success, return to userspace the encrypted key datablob size.
+ */
+static long user_encrypted_read(const struct key *key, char *buffer,
+      size_t buflen)
+{
+  struct encrypted_key_payload *epayload;
+  struct key *mkey;
+  const u8 *master_key;
+  size_t master_keylen;
+  char derived_key[HASH_SIZE];
+  char *ascii_buf;
+  size_t asciiblob_len;
+  int ret;
+
+  epayload = dereference_key_locked(key);
+
+  /* returns the hex encoded iv, encrypted-data, and hmac as ascii */
+  asciiblob_len = epayload->datablob_len + ivsize + 1
+    + roundup(epayload->decrypted_datalen, blksize)
+    + (HASH_SIZE * 2)
+    + (strcmp(epayload->format, key_format_unwrappable) ? epayload->decrypted_datalen : 0);
+
+  if (!buffer || buflen < asciiblob_len)
+    return asciiblob_len;
+
+  mkey = request_master_key(epayload, &master_key, &master_keylen);
+  if (IS_ERR(mkey))
+    return PTR_ERR(mkey);
+
+  ret = get_derived_key(derived_key, ENC_KEY, master_key, master_keylen);
+  if (ret < 0)
+    goto out;
+
+  ret = derived_key_encrypt(epayload, derived_key, sizeof derived_key);
+  if (ret < 0)
+    goto out;
+
+  ret = datablob_hmac_append(epayload, master_key, master_keylen);
```

Comment:
So this HMAC functionality will depend on the algo used.  For now I guess keep this patch just adding the new key type and then we can do a follow up with another patch for the crypto type.

Reply:
The algo modification is still up in the air, but most definitely out of scope for this patch.